File: quant1x/datasets/kline_raw.cpp
#include <quant1x/datasets/kline_raw.h>

namespace datasets {

    namespace detail {
        // 拉取数据
        std::vector<level1::SecurityBar> fetch_kline(const std::string &code, u16 start, u16 count, level1::KLineType kline_type) {
            try {
                auto conn = level1::client();
                //constexpr auto category = kline_type;
                level1::SecurityBarsRequest request(code, kline_type, start, count);
                level1::SecurityBarsResponse response(request.isIndex, kline_type);
                auto err = level1::process(conn->socket(), request, response);
                if (err) {
                    throw std::runtime_error(fmt::format("Process error: {}", err.message()));
                }
                return response.List;
            } catch (const std::exception &e) {  // 其他标准异常
                spdlog::error("[dataset::KLine] - 标准异常: {} (type: {})", e.what(), typeid(e).name());
                // 对于system_error可以记录更多信息
                if (auto se = dynamic_cast<const std::system_error *>(&e)) {
                    spdlog::error("[dataset::KLine] Error code: {}, category: {}", se->code().value(), se->code().category().name());
                }
            } catch (...) {
                spdlog::error("[dataset::KLine] 获取日K线异常");
            }
            return {};
        }
    } // namespace

    namespace {

        void save_kline_raw(const std::string &filename, const std::vector<KLineRaw>& values) {
            util::check_filepath(filename, true);
            io::CSVWriter writer(filename);
            writer.write_row("Date", "Open", "Close", "High", "Low", "Volume", "Amount", "Up", "Down", "Datetime");
            for (const auto &row: values) {
                writer.write_row(row.Date, row.Open, row.Close, row.High, row.Low, row.Volume, row.Amount,
                                 row.Up, row.Down, row.Datetime);
            }
        }
    }

    void KLineRaw::adjust(double m, double a, int number) {
        this->Open = this->Open * m + a;
        this->Close = this->Close * m + a;
        this->High = this->High * m + a;
        this->Low = this->Low * m + a;
        // 成交量复权
        // 1. 计算均价
        auto ap = this->Amount / this->Volume;
        // 2. 均价复权
        ap = ap * m + a;
        // 3. 以成交金额为基准, 用复权均价计算成交量
        this->Volume = this->Amount / ap;
        //kline->AdjustmentCount+= factor.no;
        (void)number;
    }

    std::vector<KLineRaw> read_kline_raw_from_csv(const std::string& filename) {
        std::vector<KLineRaw> klines;
        try {
            // 创建 CSV 读取器
            io::CSVReader<10> in(filename);  // 有 11 列数据

            // 设置表头字段名(用于自动匹配顺序)
            in.read_header(io::ignore_extra_column,
                           "Date", "Open", "Close", "High", "Low",
                           "Volume", "Amount", "Up", "Down", "Datetime");

            KLineRaw row = {};
            while (in.read_row(row.Date, row.Open, row.Close, row.High, row.Low, row.Volume, row.Amount,
                               row.Up, row.Down, row.Datetime)) {
                klines.emplace_back(row);
            }
        } catch(...) {
            // 忽略异常, 读csv文件失败, 返回空
        }
        return klines;
    }

    std::vector<KLineRaw> load_kline_raw(const std::string &code) {
        auto filename = config::get_kline_filename(code, false);
        spdlog::debug("[dataset::KLineRaw] kline file: {}", filename);
        return read_kline_raw_from_csv(filename);
    }

    void DataKLineRaw::Print(const std::string &code, const std::vector<exchange::timestamp> &dates)  {
        (void)code;
        (void)dates;
    }

    void DataKLineRaw::Update(const std::string &code, const exchange::timestamp &date) {
//        if(date != exchange::last_trading_day()) {
//            return;
//        }
        (void)date;
        // 1. 确定本地有效数据最后1条数据作为拉取数据的开始日期
        auto startDate = market_first_date;
        try {
            std::string cache_filename = config::get_kline_filename(code, false);
            std::vector<KLineRaw> cacheKLines = read_kline_raw_from_csv(cache_filename);
            auto kLength = cacheKLines.size();
            auto klineDaysOffset = detail::MAX_KLINE_LOOKBACK_DAYS;
            if(kLength > 0) {
                if (klineDaysOffset > kLength) {
                    klineDaysOffset = kLength;
                }
                startDate = cacheKLines[kLength-klineDaysOffset].Date;
            }
            // 2. 确定结束日期
            auto endDate = exchange::timestamp::now().pre_market_time();
            spdlog::debug("[dataset::KLineRaw] [{}]: from {} to {}", code, startDate.only_date(), endDate.only_date());
            auto ts = exchange::date_range(startDate, endDate);
            auto total = u16(ts.size());
            startDate = ts[0];
            endDate = ts[total-1];
            spdlog::debug("[dataset::KLineRaw] [{}]: from {} to {}", code, startDate.only_date(), endDate.only_date());
            u16 step = level1::security_bars_max;
            u16 start = 0;
            //u16 category = level1::RI_K;
            // 3. 拉取数据
            std::vector<std::vector<level1::SecurityBar>> hs;
            //std::vector<level1::SecurityBar> history;
            size_t elementCount = 0;
            do {
                u16 count = step;
                if(total - start >= step) {
                    count = step;
                } else {
                    count = total - start;
                }
                auto reply = detail::fetch_kline(code, start, count);
                if (reply.empty()) {
                    break;
                }
                elementCount += reply.size();
                //hs.insert(hs.end(), reply.begin(), reply.end());
                hs.emplace_back(reply);
                if (reply.size() < count) {
                    break;
                }
                start += count;
            } while (start < total);
            // 4. 由于K线数据，每次获取数据是从后往前获取, 所以这里需要反转历史数据的切片
            std::reverse(hs.begin(), hs.end());
            // 5. 调整成交量, 单位从手改成股, vol字段 * 100
            std::vector<KLineRaw> incremental_klines;
            incremental_klines.reserve(elementCount);
            for(const auto & vec : hs) {
                for (const auto & row : vec) {
                    auto dateTime = exchange::timestamp(row.Year, row.Month, row.Day).pre_market_time();
                    if (dateTime < startDate || dateTime > endDate) {
                        continue;
                    }
                    auto kx = KLineRaw{
                        .Date = dateTime.only_date(), // 日期
                        .Open = row.Open,             // 开盘价
                        .Close = row.Close,           // 收盘价
                        .High = row.High,             // 最高价
                        .Low = row.Low,               // 最低价
                        .Volume = row.Vol * 100,      // 成交量(股)
                        .Amount = row.Amount,         // 成交金额(元)
                        .Up = row.UpCount,            // 上涨家数 / 外盘
                        .Down = row.DownCount,        // 下跌家数 / 内盘
                        .Datetime = row.DateTime,     // 时间
                    };
                    incremental_klines.emplace_back(kx);
                }
            }
            // 6. K线数据转换成KLine结构
            // 6.1 判断是否已除权的依据是当前更新K线只有1条记录
//            bool adjusted = incremental_klines.size() == 1;
//            auto dividends = load_xdxr(code);
//            if (adjusted) {
//                calculate_pre_adjust(incremental_klines, startDate, dividends);
//            }
            // 6.2 只前复权当日数据
            // 7. 拼接缓存和新增的数据
            std::vector<KLineRaw> klines;
            // 7.1 先截取本地缓存的数据
            if (kLength > klineDaysOffset) {
                klines.insert(klines.end(), cacheKLines.begin(), cacheKLines.begin()+(kLength-klineDaysOffset));
            }
            // 7.2 拼接新增的数据
            if (klines.empty()) {
                klines = incremental_klines;
            } else {
                klines.insert(klines.end(), incremental_klines.begin(), incremental_klines.end());
            }
            // 8. 前复权
//            if(!adjusted) {
//                calculate_pre_adjust(klines, startDate, dividends);
//            }
            // 9. 刷新缓存文件
            save_kline_raw(cache_filename, klines);
        } catch (const std::exception &e) {  // 其他标准异常
            spdlog::error("[dataset::KLineRaw] - 标准异常: {} (type: {})", e.what(), typeid(e).name());
            // 对于system_error可以记录更多信息
            if (auto se = dynamic_cast<const std::system_error *>(&e)) {
                spdlog::error("[dataset::KLineRaw] Error code: {}, category: {}", se->code().value(), se->code().category().name());
            }
        } catch (...) {
            spdlog::error("[dataset::KLineRaw] 获取日K线异常");
        }
    }


} // namespace datasets

File: quant1x/datasets/kline_raw.go
// 单一实现（基于 C++ datasets/kline_raw.cpp）
package datasets

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"

	"gitee.com/quant1x/quant1x/quant1x/config"
	"gitee.com/quant1x/quant1x/quant1x/data"
	"gitee.com/quant1x/quant1x/quant1x/exchange"
	"gitee.com/quant1x/quant1x/quant1x/level1"
)

// KLineRaw 对应 C++ 中的 datasets::KLineRaw，用于表示原始日线数据。
type KLineRaw struct {
	Date     string  `name:"日期" csv:"date"`        // 日期 YYYY-MM-DD
	Open     float64 `name:"开盘价" csv:"open"`       // 开盘价
	Close    float64 `name:"收盘价" csv:"close"`      // 收盘价
	High     float64 `name:"最高价" csv:"high"`       // 最高价
	Low      float64 `name:"最低价" csv:"low"`        // 最低价
	Volume   float64 `name:"成交量(股)" csv:"volume"`  // 成交量(股)
	Amount   float64 `name:"成交额(元)" csv:"amount"`  // 成交额(元)
	Up       int     `name:"上涨家数 / 外盘" csv:"up"`   // 上涨家数 / 外盘
	Down     int     `name:"下跌家数 / 内盘" csv:"down"` // 下跌家数 / 内盘
	Datetime string  `name:"时间" csv:"datetime"`    // 时间
}

const maxKlineRawLookbackDays = 1

// FetchKLineRaw 从 level1 拉取原始 K 线数据并转换为 datasets.KLineRaw。
func FetchKLineRaw(securityCode string, start, count uint16, klineType level1.KLineType) ([]level1.SecurityBar, error) {
	conn, release, err := level1.GetStdConnection()
	if err != nil {
		return nil, fmt.Errorf("level1 client acquire failed: %w", err)
	}
	if release != nil {
		defer release()
	}
	if conn == nil || conn.Conn() == nil {
		return nil, fmt.Errorf("nil connection from level1 client")
	}

	req := level1.NewSecurityBarsRequest(securityCode, klineType, start, count)
	resp := level1.NewSecurityBarsResponse(req.IsIndex, uint16(req.Param.Category))
	if err := level1.Process(conn, req, resp); err != nil {
		return nil, fmt.Errorf("security bars request failed: %w", err)
	}

	return resp.List, nil
}

// SaveKlineRaw 将原始 K 线写入 CSV 文件（包含表头）。
func SaveKlineRaw(filename string, values []KLineRaw) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()

	header := []string{"date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime"}
	if err := w.Write(header); err != nil {
		return err
	}
	for _, r := range values {
		row := []string{
			r.Date,
			fmt.Sprintf("%g", r.Open),
			fmt.Sprintf("%g", r.Close),
			fmt.Sprintf("%g", r.High),
			fmt.Sprintf("%g", r.Low),
			fmt.Sprintf("%g", r.Volume),
			fmt.Sprintf("%g", r.Amount),
			fmt.Sprintf("%d", r.Up),
			fmt.Sprintf("%d", r.Down),
			r.Datetime,
		}
		if err := w.Write(row); err != nil {
			return err
		}
	}
	return nil
}

// ReadKlineRawFromCSV 从 CSV 文件中读取 KLineRaw 条目；解析错误会返回错误。
func ReadKlineRawFromCSV(filename string) ([]KLineRaw, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	r := csv.NewReader(f)
	rows, err := r.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}
	// 期望第一行为表头
	out := make([]KLineRaw, 0, len(rows)-1)
	for i := 1; i < len(rows); i++ {
		rec := rows[i]
		if len(rec) < 10 {
			continue
		}
		open, _ := strconv.ParseFloat(rec[1], 64)
		closev, _ := strconv.ParseFloat(rec[2], 64)
		high, _ := strconv.ParseFloat(rec[3], 64)
		low, _ := strconv.ParseFloat(rec[4], 64)
		vol, _ := strconv.ParseFloat(rec[5], 64)
		amt, _ := strconv.ParseFloat(rec[6], 64)
		up, _ := strconv.Atoi(rec[7])
		down, _ := strconv.Atoi(rec[8])
		datetime := rec[9]
		k := KLineRaw{
			Date:     rec[0],
			Open:     open,
			Close:    closev,
			High:     high,
			Low:      low,
			Volume:   vol,
			Amount:   amt,
			Up:       up,
			Down:     down,
			Datetime: datetime,
		}
		out = append(out, k)
	}
	return out, nil
}

// LoadKlineRaw 从缓存文件加载原始K线数据。
func LoadKlineRaw(code string) ([]KLineRaw, error) {
	filename := config.GetKlineFilename(code, false)
	log.Printf("[dataset::KLineRaw] kline file: %s", filename)
	return ReadKlineRawFromCSV(filename)
}

// DataKLineRaw 实现了与 C++ DataKLineRaw 类似的缓存适配器更新逻辑
type DataKLineRaw struct{}

// 实现 data.DataAdapter 的 Schema 方法
func (d *DataKLineRaw) Kind() data.Kind { return BaseRawDailyKLine }
func (d *DataKLineRaw) Owner() string   { return data.DefaultDataProvider }
func (d *DataKLineRaw) Key() string     { return "day_raw" }
func (d *DataKLineRaw) Name() string    { return "日K线RAW" }
func (d *DataKLineRaw) Usage() string   { return "日K线RAW" }

// Print 实现 data.DataAdapter.Print（可变参数日期）
func (d *DataKLineRaw) Print(code string, dates ...exchange.Timestamp) {
	_ = code
	_ = dates
}

// Update 对应 C++ DataKLineRaw::Update 的行为：读取本地缓存、确定时间窗口、分页拉取 level1 数据、
// 反转与合并结果，并写回缓存文件。
func (d *DataKLineRaw) Update(code string, _date exchange.Timestamp) {
	// 1. 确定缓存文件并读取本地缓存
	cacheFilename := config.GetKlineFilename(code, false)
	cacheKLines, _ := ReadKlineRawFromCSV(cacheFilename) // ignore parse errors, follow C++ behavior
	klinesLength := len(cacheKLines)
	klinesOffsetDays := maxKlineRawLookbackDays

	// 默认起始日期（使用 datasets.MarketFirstDate，与 C++ 的 market_first_date 等价）
	currentStartDate := MarketFirstDate
	if klinesLength > 0 {
		if klinesOffsetDays > klinesLength {
			klinesOffsetDays = klinesLength
		}
		kline := cacheKLines[klinesLength-klinesOffsetDays]
		if ts, err := exchange.ParseTimestamp(kline.Date); err == nil {
			currentStartDate = ts
		}
	}

	// 2. 确定结束日期
	currentEndDate := exchange.NowTimestamp().PreMarketTime()
	log.Printf("[dataset::KLineRaw] [%s]: from %s to %s", code, currentStartDate.OnlyDate(), currentEndDate.OnlyDate())

	step := uint16(level1.SecurityBarsMax)
	start := uint16(0)
	hs := make([][]level1.SecurityBar, 0)
	elementCount := 0

	for {
		count := step
		reply, err := FetchKLineRaw(code, start, uint16(count), level1.KLineDaily)
		if err != nil || len(reply) == 0 {
			break
		}

		elementCount += len(reply)
		hs = append(hs, reply)

		lastBar := reply[len(reply)-1]
		lastBarDate := exchange.PreMarketTimestamp(lastBar.Year, lastBar.Month, lastBar.Day)

		if lastBarDate.Less(currentStartDate) {
			break
		}

		if len(reply) < int(count) {
			break
		}

		start += uint16(count)
	}

	// 反转切片以获得正确的顺序
	for i, j := 0, len(hs)-1; i < j; i, j = i+1, j-1 {
		hs[i], hs[j] = hs[j], hs[i]
	}

	incrementalKLines := make([]KLineRaw, 0)

	for _, vec := range hs {
		for _, row := range vec {
			dateTime := exchange.PreMarketTimestamp(row.Year, row.Month, row.Day)

			if dateTime.Less(currentStartDate) || dateTime.Greater(currentEndDate) {
				continue
			}

			kx := KLineRaw{
				Date:     dateTime.OnlyDate(),
				Open:     row.Open,
				Close:    row.Close,
				High:     row.High,
				Low:      row.Low,
				Volume:   row.Vol * 100, // Convert to shares
				Amount:   row.Amount,
				Up:       int(row.UpCount),
				Down:     int(row.DownCount),
				Datetime: row.DateTime,
			}
			incrementalKLines = append(incrementalKLines, kx)
		}
	}

	// 7. 合并
	klines := make([]KLineRaw, 0)
	if klinesLength > klinesOffsetDays {
		klines = append(klines, cacheKLines[:klinesLength-klinesOffsetDays]...)
	}
	klines = append(klines, incrementalKLines...)

	// 9. 保存
	if err := SaveKlineRaw(cacheFilename, klines); err != nil {
		log.Printf("[dataset::KLineRaw] save error: %v", err)
	}
}

func init() {
	// 注册DataKLineRaw插件
	if err := data.Register(&DataKLineRaw{}); err != nil {
		log.Printf("[dataset::KLineRaw] failed to register plugin: %v", err)
	}
}
